packages/turbo/bin/turbo

```diff
@@ -1,10 +1,257 @@
+  const env = { ...process.env, npm_config_global: undefined };
+
+  child_process.execSync(
+    `npm install --loglevel=error --prefer-offline --no-audit --progress=false`,
+    { cwd: turboPath, stdio: "pipe", env }
+  );
+}
+
+// This provides logging messages as it progresses towards calculating the binary path.
+function getBinaryPath() {
+  // First we see if the user has configured a particular binary path.
+  const TURBO_BINARY_PATH = process.env.TURBO_BINARY_PATH;
+  if (TURBO_BINARY_PATH) {
+    if (!fs.existsSync(TURBO_BINARY_PATH)) {
+      console.warn(
+        `[turbo] Ignoring bad configuration: TURBO_BINARY_PATH=${TURBO_BINARY_PATH}`
+      );
+    } else {
+      return TURBO_BINARY_PATH;
+    }
+  }
+
+  const availablePlatforms = [
+    'darwin',
+    'linux',
+    'windows',
+  ];
+
+  const availableArchs = [
+    '64',
+    'arm64',
+  ];
+
+  // We need to figure out which binary to hand the user.
+  // The only place where the binary can be at this point is `require.resolve`-able
+  // relative to this package as it should be installed as an optional dependency.
+
+  const { platform, arch } = process;
+  const resolvedArch = arch === 'x64' ? '64' : arch;
+  const ext = platform === 'windows' ? '.exe' : '';
+
+  // Try all places in order until we get a hit.
+
+  // 1. The package which contains the binary we _should_ be running.
+  const correctBinary = availablePlatforms.includes(platform) && availableArchs.includes(resolvedArch) ? `turbo-${platform}-${resolvedArch}/bin/turbo${ext}` : null;
+  if (correctBinary !== null) {
+    try {
+      return require.resolve(`${correctBinary}`);
+    } catch (e) {}
+  }
+
+  // 2. Install the binary that they need just in time.
+  if (SHOULD_INSTALL && correctBinary !== null) {
+    console.warn('Turborepo did not find the correct binary for your platform.');
+    console.warn('We will attempt to install it now.');
+
+    try {
+      installUsingNPM();
+      const resolvedPath = require.resolve(`${correctBinary}`);
+      console.warn('Installation has succeeded.');
+      return resolvedPath;
+      } catch (e) {
+      console.warn('Installation has failed.');
+    }
+  }
+
+  // 3. Both Windows and macOS ARM boxes can run x64 binaries. Attempt to run under emulation.
+  const alternateBinary = (arch === "arm64" && ['darwin', 'windows'].includes(platform)) ? `turbo-${platform}-64/bin/turbo${ext}` : null;
+  if (SHOULD_ATTEMPT_EMULATED && alternateBinary !== null) {
+    try {
+      const resolvedPath = require.resolve(`${alternateBinary}`);
+      console.warn(`Turborepo detected that you're running:\n${platform} ${resolvedArch}.`);
+      console.warn(`We were not able to find the binary at:\n${correctBinary}`);
+      console.warn(`We found a possibly-compatible binary at:\n${alternateBinary}`);
+      console.warn(`We will attempt to run that binary.`);
+      return resolvedPath;
+    } catch (e) {}
+  }
+
+  // We are not going to run `turbo` this invocation.
```

Comment:
[You will not go to space today.](https://www.explainxkcd.com/wiki/index.php/1133:_Up_Goer_Five)